src/libs/elektra/opmphm_vheap.c

```diff
@@ -0,0 +1,141 @@
+#include "kdbprivate.h"
+
+/**
+ * Allocates vheap with size specified by parameter minSize
+ *
+ * To construct a max heap the comparison function VheapComp (a, b), must return
+ * 1 on a > b and 0 otherwise. For a min heap 1 on a < b and 0 otherwise.
+ *
+ * @ingroup vheap
+ * @param comp the comparison function of the heap
+ * @param minSize the minimum size of the heap
+ * @return a Vheap pointer
+ * @return NULL error
+ */
+Vheap * elektraVheapInit (VheapComp comp, size_t minSize)
+{
+	if (minSize < 1 || !comp) return NULL;
+	Vheap * newVheap = elektraMalloc (sizeof (Vheap));
+	if (!newVheap) return NULL;
+	newVheap->comp = comp;
+	newVheap->size = minSize;
+	newVheap->minSize = minSize;
+	newVheap->count = 0;
+	newVheap->data = elektraMalloc (newVheap->size * sizeof (void *));
+	if (!newVheap->data)
+	{
+		elektraFree (newVheap);
+		return NULL;
+	}
+	return newVheap;
+}
+
+/**
+ * Checks if the heap is empty.
+ *
+ * @ingroup vheap
+ * @return 1 on empty
+ * @return 0 on non empty
+ */
+int elektraVheapIsEmpty (const Vheap * vheap)
+{
+	if (!vheap) return 0;
+	return (vheap->count == 0) ? 1 : 0;
+}
+
+/**
+ * Deletes the heap, by freeing all the memory.
+ * @ingroup vheap
+ *
+ */
+void elektraVheapDel (Vheap * vheap)
+{
+	if (!vheap) return;
+	elektraFree (vheap->data);
+	elektraFree (vheap);
+}
+
+/**
+ * Inserts an element in the Vheap, by finding the right position.
+ * Resizes the memory first if needed.
+ *
+ * @ingroup vheap
+ * @param data the element
+ * @return 0 on error
+ * @return 1 otherwise
+ */
+int elektraVheapInsert (Vheap * vheap, void * data)
+{
+	if (!vheap) return 0;
+	// grow
+	if (++vheap->count > vheap->size)
+	{
+		vheap->size <<= 1;
+		if (elektraRealloc ((void **)&vheap->data, vheap->size * sizeof (void *)) == -1)
+		{
+			return 0;
```

Comment:
I'm not sure if this is really an issue, but in line 71 the heap count is incremented and not decremented (i.e. rollback) here before returning